### DISCUSSION TITLE
false colors in ULA+ mode?

### DISCUSSION THREAD

**Discussion by tuFFo126** [UzixLS/zx-sizif-512#44] [General]
The ULA+ system in sizif clearly distorts colors.  The left screen is the image from sizif, the right screen is the image from zx-hd.
The colors of the images in sizif seem unnatural.
While standard zx colors have no distortions.
By running the "testrgb.tap" program from the set of programs for ula+, i noticed that for green the color does not change when going from value 3 to 4, the same is true for red.
![20231004_220552](https://github.com/UzixLS/zx-sizif-512/assets/124886433/7c48e3ac-1c03-46ae-905e-2c6bc3febc0b)
![20231004_220823](https://github.com/UzixLS/zx-sizif-512/assets/124886433/79ae9fef-ecec-4c60-a218-e981c4ab7302)
![20231004_220759](https://github.com/UzixLS/zx-sizif-512/assets/124886433/d914da97-2b8f-4324-ab82-22e3a45aa6ad)
![20231004_221948](https://github.com/UzixLS/zx-sizif-512/assets/124886433/ab3c804e-3fd3-4df4-be35-5b8d420f4672)

**Comment by UzixLS** [comment]
Hmm, ULA+ standard assume 8-bit color, but Sizif implements only 6-bit color in hardware and trying to achieve missing colors using well-known PWM trick - by fast switching between two different colors. Usually that produces decent result, but your test clearly show imperfection. I believe there is room for improvements in current PWM algorithm.
6-bit color have historical reason as Sizif initially doesn't designed with ULA+ in mind, but only Red + RedIntesity + Green + GreenIntensity + Blue + BlueIntensity standard Spectrum palette.

**Comment by UzixLS** [comment]
Can you share your test images? I want to check them on my board

**Reply by tuFFo126** [reply]
everything is taken from the original ULA+ project website
https://sourcesolutions.itch.io/ulaplus10

**Reply by tuFFo126** [reply]
Were you able to download and check it?

**Reply by UzixLS** [reply]
Still haven't had time for this, sorry

**Reply by tuFFo126** [reply]
Ok, I'll wait, but I'm curious if it's just a problem with my copy or the design?

**Reply by tuFFo126** [reply]
Were you able to check anything?

**Reply by tuFFo126** [reply]
I wonder if, apart from uploading the appropriate version of the input to the altera, other elements also have an impact? See discussion. Because it looks like in his sizif the colors of the ula+ are correct.  I asked for a check it.
https://www.facebook.com/share/p/ouQc5fEMyZrdBqFM/
